maven/impl-maven-embedded/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/embedded/MavenDownloadsTestCase.java

```diff
@@ -56,6 +65,60 @@ public void testDownloadFromGivenDestinationWithoutUsingCacheAndExtract() throws
         verifyExtraction(2, "resolver-3.0.0-alpha-1", "resolver-3.0.0-alpha-3");
     }
 
+    @Test
+    public void testDownloadInMultipleThreads() throws IOException, InterruptedException {
+        // cleanup
+        FileUtils.deleteDirectory(targetMavenDir);
+
+        // multiple download
+        CountDownLatch latch = new CountDownLatch(1);
+        final Thread firstDownload = createThreadWithDownload(latch);
+        final Thread secondDownload = createThreadWithDownload(latch);
+        firstDownload.start();
+        secondDownload.start();
+        latch.countDown();
+        Awaitility.await().until(new Callable<Boolean>() {
+            @Override
+            public Boolean call() throws Exception {
+                return !firstDownload.isAlive() && !secondDownload.isAlive();
+            }
+        });
+
+        // verify
+        String expMsg = "Resolver: downloading Maven binaries from";
+        Matcher matcher = Pattern.compile(expMsg).matcher(systemOutRule.getLog());
+        assertThat(matcher.find()).as(String.format(
+                "The log should contain one occurrence of message \"%s\" but none was found. For more information see the log", expMsg))
+                                  .isTrue();
+        assertThat(matcher.find()).as(String.format(
+                "The log should contain only one occurrence of message \"%s\" but more than one was found. For more information see the log",
+                expMsg)).isFalse();
+    }
+
+    private Thread createThreadWithDownload(final CountDownLatch latch) {
+        return new Thread(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    latch.await();
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                }
+                downloadSWR300Alpha1();
+            }
+        });
+    }
+
+    private void downloadSWR300Alpha1() {
+        try {
+            EmbeddedMaven
+                    .forProject(pathToJarSamplePom)
+                    .useDistribution(new URL("http://github.com/shrinkwrap/resolver/archive/3.0.0-alpha-1.zip"), false);
```

Comment:
I was also wondering - doesn't it make this test vulnerable for network issues? Could we somehow simulate download process instead so we have it all locally?